### PR TITLE
iox-#590 Separate errors of posh and C binding

### LINKED ISSUE #590: Improve modularity by creating architecure guidelines
## Brief feature description

iceoryx is developed in several CMake packages that aim to be independent. Current state is there are cross dependencies between each of them and no architecture guidelines have been created so far.

## Detailed information

Analysed iceoryx repository with [Lattix Architect](https://www.lattix.com/products-architecture-issues/). [Attached](https://github.com/eclipse-iceoryx/iceoryx/files/6071770/iceoryx_b929aa_Architecture.Metrics_Metrics.Architecture.xlsx) are the the dependecy structure matrix and the architecture metrixs of Lattix (commit b929aa8d5022471a318cdc8e6994f1cdd6aa3850).

### General

* [x] Analyse Lattix
  * [x] Dependecy structure matrix (DSM)
  * [x] Architectual metrics
  * [x] Conceptual architecture
* [x] Namespaces & CMake package/library refactoring
* [x] Add CMake target for quality level 5/playground features and move "maybe obsolote" constructs there
  * [x] Add script to compare used namespaces to a whitelist in binary (`nm -C myBinary -f posix` with C++filt for demangling) to make sure no experimental/playground code is baked into releases
* [ ] RouDi modularisation #539
* [x] Create architecture documentation
  * [x] PlantUML or plain ASCII art? 
  * [x] Package view
  * [x] Layer view
* [ ] Summarise guidelines and write a CI checker
  * [ ] Remove `iox::helper` namespace
* [ ] Create `iceoryx_posh_core` CMake target and address the `@todo #590` in the CMake files
* [ ] Remove unused files and move quality level 2 files to a separate CMake package

### Lattix Architect findings

* [x] ~~Move POSH specific errors from iceoryx_utils to iceoryx_posh~~ Moved to #1099 
* [x] `roudi_gtest.hpp` includes `iceoryx_posh/internal/roudi_environment/roudi_environment.hpp`
* [ ] Move `RouDiEnvironment` to test folder & namespace
* [ ] `iceoryx_hoofs` does not have a clear architecture and shall be better structured

### PR DESCRIPTION
**Pull request by mossmaurice**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [ ] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CHANGELOG.md

* Move errors from hoofs to specific module (posh/ C binding) to make hoofs independent from posh and C binding

### Todo
* Swap arguments in `errorHandler()` call so that callable is last to avoid `nullptr` arg

### Discussion
* Switch to a builder pattern as next step?
* How to make the enum values unique?
  * Use an additional component identifier in the `errorHandler()` call? 
* Remove the `k` in front of the errors?

## Checklist for the PR Reviewer

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] Code according to our coding style and naming conventions
- [ ] Unit tests have been written for new behavior
    - [ ] Each unit test case has a unique UUID
- [ ] Public API changes are documented via doxygen
- [ ] Copyright owner are updated in the changed files
- [ ] PR title describes the changes

## Post-review Checklist for the PR Author

1. [ ] All open points are addressed and tracked via issues

## References

- Closes #590

### CODE REVIEW

File: iceoryx_binding_c/include/iceoryx_binding_c/error_handling/error_handling.hpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    error(BINDING_C__PUBLISHER_OPTIONS_NOT_INITIALIZED) \
+    error(BINDING_C__SUBSCRIBER_OPTIONS_NOT_INITIALIZED) \
+    error(BINDING_C__C2CPP_ENUM_TRANSLATION_INVALID_SUBSCRIBER_EVENT_VALUE) \
+    error(BINDING_C__C2CPP_ENUM_TRANSLATION_INVALID_SUBSCRIBER_STATE_VALUE)
+
+// clang-format on
+
+enum class CBindingError : uint32_t
+{
+    kNO_ERROR = 2000,
```
Comment (mossmaurice) [inline iceoryx_binding_c/include/iceoryx_binding_c/error_handling/error_handling.hpp:37]: This is one solution to the question "How to make the enum values unique all over iceoryx?" Another possibility would be to add an additional enum to the call with a component id.

**Review by elfenpiff — CHANGES_REQUESTED** [review]
Can we please keep this PR on hold until we figured out how the overall error concept in iceoryx looks like.
To call terminate is most likely not an option for ASIL D and we may have to make this either configurable or that some user can define custom callbacks to notify maybe some kind of watchdog or even throw an exception.

In those cases an error handler may is the more viable solution.

**Review by elfenpiff — CHANGES_REQUESTED** [review]
You checked the PR title check but the title does not follow our guidelines. Please add `iox-#???`.

**Review by elfenpiff — CHANGES_REQUESTED** [review]
Please perform the renaming of `error_handling.hpp` into `error_handler.hpp` in a separate PR.

**Comment by mossmaurice** [comment]
This PR will be rebased onto #1100. Please refrain from reviewing it. I'll leave it open until there is a replacement to link to.

**Comment by mossmaurice** [comment]
Superseded by #1170

**Pull request closed without merging**